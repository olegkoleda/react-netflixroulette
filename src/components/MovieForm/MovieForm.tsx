import React, { useCallback, useMemo, useState } from "react";
import Input from "../Input";
import Select from "../Select";
import Button from "../Button";
import { StyledLabel, StyledText } from "../Input/styled.input";
import { useInput } from "../../hooks/useInput";
import { Box } from "atomic-layout";
import { IMovie } from "../../interfaces/IMovie";

interface IFormProps {
  movieData?: IMovie;
  callback: Function;
}
const options = [
  { label: "Drama", value: "Drama" },
  { label: "Romance", value: "Romance" },
  { label: "Animation", value: "Animation" },
  { label: "Adventure", value: "Adventure" },
  { label: "Family", value: "Family" },
  { label: "Comedy", value: "Comedy" },
  { label: "Fantasy", value: "Fantasy" },
  { label: "Science Fiction", value: "Science Fiction" },
  { label: "Action", value: "Action" },
  { label: "Mystery", value: "Mystery" },
  { label: "Thriller", value: "Thriller" },
  { label: "Music", value: "Music" },
  { label: "War", value: "War" },
  { label: "Crime", value: "Crime" },
  { label: "History", value: "History" },
  { label: "Horror", value: "Horror" },
  { label: "Western", value: "Western" },
  { label: "Documentary", value: "Documentary" },
  { label: "TV Movie", value: "TV Movie" },
];

const MovieForm: React.FC<IFormProps> = ({ movieData, callback }) => {
  const selectedGenres =
    useMemo(() => movieData?.genres.map((value) => ({ label: value, value })), [
      movieData,
    ]) || [];

  const [selected, setSelected] = useState(selectedGenres);
  const { value: title, bind: bindTitle, reset: resetTitle } = useInput(
    movieData?.title || ""
  );
  const { value: date, bind: bindDate, reset: resetDate } = useInput(
    movieData?.release_date || ""
  );
  const { value: url, bind: bindUrl, reset: resetUrl } = useInput(
    movieData?.poster_path || ""
  );
  const { value: runtime, bind: bindRuntime, reset: resetRuntime } = useInput(
    "" + movieData?.runtime || ""
  );
  const {
    value: overview,
    bind: bindOverview,
    reset: resetOverview,
  } = useInput(movieData?.overview || "");

  const genres = useMemo(() => selected.map(({ value }) => value), [selected]);

  const handleSubmit = useCallback(
    (e) => {
      e.preventDefault();
      const formData = {
        ...movieData,
        title,
        release_date: date,
        poster_path: url,
        overview,
        runtime: +runtime,
        genres,
      };

      callback(formData);
    },
    [title, date, url, overview, runtime, genres]
  );

  const handleReset = useCallback(
    (e) => {
      e.preventDefault();
      resetTitle();
      resetDate();
      resetUrl();
      resetRuntime();
      resetOverview();
      setSelected([]);
    },
    [resetTitle, resetDate, resetUrl, resetRuntime, resetOverview, setSelected]
  );

  return (
    <form onSubmit={handleSubmit}>
      {movieData?.id ? (
        <StyledLabel>
          {"Movie ID"}
          <StyledText>{movieData?.id}</StyledText>
        </StyledLabel>
      ) : (
        false
      )}
      <Input label="Title" placeholder="Select Title" {...bindTitle} />
      <Input
        type="date"
        label="Release date"
        placeholder="Select Date"
        {...bindDate}
      />
      <Input label="movie URL" placeholder="Movie URL here" {...bindUrl} />
      <Select
        options={options}
        selected={selected}
        setSelected={setSelected}
        label={"Genre"}
        placeholder={"Select Genre"}
        search={true}
      />
      <Input label="Overview" placeholder="Overview here" {...bindOverview} />
      <Input
        label="Runtime"
        placeholder="Runtime here"
        type={"number"}
        {...bindRuntime}
      />
      <Box flex justifyContent={"space-between"} marginTop={"2rem"}>
        <Button secondary onClick={handleReset}>
          Reset
        </Button>
        <Button>Submit</Button>
      </Box>
    </form>
  );
};

export default MovieForm;
