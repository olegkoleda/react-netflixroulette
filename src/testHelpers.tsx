import React from "react";
import renderer from "react-test-renderer";
import { theme } from "./styles/Theme";
import { ThemeProvider } from "styled-components";

export function renderWithTheme(component: React.ReactNode) {
  return renderer.create(
    <ThemeProvider theme={theme}>{component}</ThemeProvider>
  );
}
