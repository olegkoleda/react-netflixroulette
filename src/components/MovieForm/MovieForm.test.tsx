import React from "react";
import MovieForm from "./MovieForm";
import { renderWithTheme } from "../../testHelpers";

const callback = jest.fn();

test("MovieForm component matches the Snapshot", () => {
  const component = renderWithTheme(<MovieForm submitCallback={callback} />);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
