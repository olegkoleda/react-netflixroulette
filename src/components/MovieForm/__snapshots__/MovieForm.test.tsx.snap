// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieForm component matches the Snapshot 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="styledinput__StyledInputWrapper-sc-1rvx5si-3 kInCGe"
  >
    <label
      className="styledinput__StyledLabel-sc-1rvx5si-1 cNeYRS"
    >
      Title
      <input
        className="styledinput__StyledInput-sc-1rvx5si-0 kqVAIk"
        name="title"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Select Title"
        value=""
      />
    </label>
  </div>
  <div
    className="styledinput__StyledInputWrapper-sc-1rvx5si-3 kInCGe"
  >
    <label
      className="styledinput__StyledLabel-sc-1rvx5si-1 cNeYRS"
    >
      Release date
      <input
        className="styledinput__StyledInput-sc-1rvx5si-0 kqVAIk"
        name="release_date"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Select Date"
        type="date"
        value=""
      />
    </label>
  </div>
  <div
    className="styledinput__StyledInputWrapper-sc-1rvx5si-3 kInCGe"
  >
    <label
      className="styledinput__StyledLabel-sc-1rvx5si-1 cNeYRS"
    >
      movie URL
      <input
        className="styledinput__StyledInput-sc-1rvx5si-0 kqVAIk"
        name="poster_path"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Movie URL here"
        value=""
      />
    </label>
  </div>
  <div
    className="styledinput__StyledInputWrapper-sc-1rvx5si-3 kInCGe"
  >
    <label
      className="styledinput__StyledLabel-sc-1rvx5si-1 cNeYRS"
    >
      <div
        className="styledselect__StyledSelect-sc-36t145-0 bHPDPW"
        onClick={[Function]}
      >
        Genre
        <div
          className="go2139093995 multi-select"
        >
          <div
            aria-expanded={false}
            aria-labelledby="Select Genre"
            aria-readonly={true}
            className="go2646822163 dropdown-container"
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            tabIndex={0}
          >
            <div
              className="go2642161244 dropdown-heading"
              onClick={[Function]}
            >
              <div
                className="dropdown-heading-value"
              >
                <span
                  className="gray"
                >
                  Select...
                </span>
              </div>
              <svg
                className="dropdown-heading-dropdown-arrow gray"
                fill="none"
                height="24"
                stroke="currentColor"
                strokeWidth="2"
                width="24"
              >
                <path
                  d="M6 9L12 15 18 9"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </label>
  </div>
  <div
    className="styledinput__StyledInputWrapper-sc-1rvx5si-3 kInCGe"
  >
    <label
      className="styledinput__StyledLabel-sc-1rvx5si-1 cNeYRS"
    >
      Overview
      <input
        className="styledinput__StyledInput-sc-1rvx5si-0 kqVAIk"
        name="overview"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Overview here"
        value=""
      />
    </label>
  </div>
  <div
    className="styledinput__StyledInputWrapper-sc-1rvx5si-3 kInCGe"
  >
    <label
      className="styledinput__StyledLabel-sc-1rvx5si-1 cNeYRS"
    >
      Runtime
      <input
        className="styledinput__StyledInput-sc-1rvx5si-0 kqVAIk"
        name="runtime"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Runtime here"
        type="number"
        value=""
      />
    </label>
  </div>
  <div
    className="sc-AxjAm jBojTN"
  >
    <button
      className="styledbutton__StyledButton-sc-7hz9n2-0 lnvFAg"
      onClick={[Function]}
      type="button"
    >
      Reset
    </button>
    <button
      className="styledbutton__StyledButton-sc-7hz9n2-0 dUEACD"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
